include/onnxruntime/core/framework/data_types.h

```diff
@@ -497,7 +550,81 @@ class SparseTensorType : public SparseTensorTypeBase {
  private:
   SparseTensorType() {
     using namespace data_types_internal;
-    TensorElementTypeSetter<elemT>::SetSparseTensorElementType(mutable_type_proto());
+    TensorElementTypeSetter<elemT>::SetSparseTensorElementType(this->mutable_type_proto());
+  }
+};
+
+/// Common base-class for all optional types.
+class OptionalTypeBase : public DataTypeImpl {
+ public:
+  static MLDataType Type();
+
+  bool IsOptionalType() const override {
+    return true;
+  }
+
+  const OptionalTypeBase* AsOptionalType() const override {
+    return this;
+  }
+
+  bool IsCompatible(const ONNX_NAMESPACE::TypeProto& type_proto) const override;
+
+  size_t Size() const override {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  DeleteFunc GetDeleteFunc() const override {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  const ONNX_NAMESPACE::TypeProto* GetTypeProto() const override;
+
+  virtual MLDataType GetElementType() const {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  OptionalTypeBase(const OptionalTypeBase&) = delete;
+  OptionalTypeBase& operator=(const OptionalTypeBase&) = delete;
+
+ protected:
+  ONNX_NAMESPACE::TypeProto& mutable_type_proto();
+
```

Comment:
MutableTypeProto?

Reply:
Needed fixing across data_types.h/data_types.cc - Fixed it.